modules/apps/questions/questions-web/src/main/java/com/liferay/questions/web/internal/upgrade/v1_1_0/QuestionsConfigurationUpgradeProcess.java

```diff
@@ -0,0 +1,145 @@
+import com.liferay.questions.web.internal.configuration.QuestionsConfiguration;
+
+import java.util.Dictionary;
+
+import org.osgi.framework.Constants;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+
+/**
+ * @author Marco Galluzzi
+ */
+public class QuestionsConfigurationUpgradeProcess extends UpgradeProcess {
+
+	public QuestionsConfigurationUpgradeProcess(
+		ConfigurationAdmin configurationAdmin,
+		ConfigurationProvider configurationProvider,
+		MBCategoryLocalService mbCategoryLocalService) {
+
+		_configurationAdmin = configurationAdmin;
+		_configurationProvider = configurationProvider;
+		_mbCategoryLocalService = mbCategoryLocalService;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		_updateCompanyConfigurations();
+		_updateSystemConfiguration();
+	}
+
+	private Configuration[] _getScopedConfigurations() throws Exception {
+		Configuration[] configurations = _configurationAdmin.listConfigurations(
+			String.format(
+				"(%s=%s.scoped)", ConfigurationAdmin.SERVICE_FACTORYPID,
+				_CLASS_NAME_QUESTIONS_CONFIGURATION));
+
+		if (configurations == null) {
+			return new Configuration[0];
+		}
+
+		return configurations;
+	}
+
+	private Configuration _getSystemConfiguration() throws Exception {
+		Configuration[] configurations = _configurationAdmin.listConfigurations(
+			String.format(
+				"(%s=%s)", Constants.SERVICE_PID,
+				_CLASS_NAME_QUESTIONS_CONFIGURATION));
+
+		if (configurations == null) {
+			return null;
+		}
+
+		return configurations[0];
+	}
+
+	private Dictionary<String, Object> _getUpdatedProperties(
+		Configuration configuration) {
+
+		if (configuration == null) {
+			return null;
+		}
+
+		Dictionary<String, Object> properties = configuration.getProperties();
+
+		if (properties == null) {
+			return null;
+		}
+
+		String rootTopicExternalReferenceCode = StringPool.BLANK;
+		long rootTopicId = (long)properties.get("rootTopicId");
+
+		if (rootTopicId != MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID) {
+			MBCategory mbCategory = _mbCategoryLocalService.fetchMBCategory(
+				rootTopicId);
+
+			if (mbCategory != null) {
+				rootTopicExternalReferenceCode =
+					mbCategory.getExternalReferenceCode();
+			}
+		}
+
+		properties.put(
+			"rootTopicExternalReferenceCode", rootTopicExternalReferenceCode);
+
+		return properties;
+	}
+
+	private void _updateCompanyConfigurations() throws Exception {
+		Configuration[] configurations = _getScopedConfigurations();
+
+		for (Configuration configuration : configurations) {
+			Dictionary<String, Object> properties = _getUpdatedProperties(
+				configuration);
+
+			if (properties == null) {
+				continue;
+			}
+
+			long companyId = (long)properties.get(
+				ExtendedObjectClassDefinition.Scope.COMPANY.getPropertyKey());
+
+			if (companyId != 0) {
+				_configurationProvider.saveCompanyConfiguration(
+					QuestionsConfiguration.class, companyId, properties);
+			}
```

Comment:
Can you make the companyId check inside `_getUpdatedProperties`? If `companyId != 0`, then return `null`. That will make `_updateCompanyConfigurations` and `_updateSystemConfigurations` symmetrical.

Reply:
If I do that we need to:

* Pass the scope to `_getUpdatedProperties`, to check the `companyId` only in when called from `_updateCompanyConfigurations`
* Get the `companyId` from the `properties` a second time inside `_updateCompanyConfigurations` because is needed by the `saveCompanyConfiguration` call.

Do you think it is worth?

What I can change is:

```java
			if (companyId == 0) {
				continue;
			}

			_configurationProvider.saveCompanyConfiguration(
				QuestionsConfiguration.class, companyId, properties);
```

Reply:
Not worth it. Nevermind.

Reply:
Done. See last commit [574c892](https://github.com/liferay-content-management/liferay-portal/pull/5927/commits/574c892318435a0934d4a9a824b8ecb6d522ebb8)